servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1783,6 +1806,166 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
+        }
+      }
+      setAclInternal(targetInode, opTimeMs, user, null, (short) -1);
+    }
+    return true;
+  }
+
+  /**
+   * Sets the group of the path. Only a super user or the path owner can change.
+   *
+   * @param path to be set group on
+   * @param group to be set
+   * @param recursive indicates whether to set all the children recursively when path is a dir
+   * @return true if set successfully, false otherwise
+   * @throws TachyonException
+   */
+  private boolean setGroup(TachyonURI path, String group, boolean recursive) throws
+      TachyonException {
+    synchronized (mInodeTree) {
+      checkOwner(path);
+      long opTimeMs = System.currentTimeMillis();
+      Inode targetInode = mInodeTree.getInodeByPath(path);
+      if (recursive && targetInode.isDirectory()) {
+        List<Inode> inodeChildren =
+            mInodeTree.getInodeChildrenRecursive((InodeDirectory) targetInode);
+        for (Inode inode : inodeChildren) {
+          checkOwner(mInodeTree.getPath(inode));
+        }
+        for (Inode inode : inodeChildren) {
+          setAclInternal(inode, opTimeMs, null, group, (short) -1);
+        }
+      }
+      setAclInternal(targetInode, opTimeMs, null, group, (short) -1);
+    }
+    return true;
+  }
+
+  /**
+   * Sets the permission of the path. Only a super user or the path owner can change.
+   *
+   * @param path to be set permission on
+   * @param permission to be set
+   * @param recursive indicates whether to set all the children recursively when path is a dir
+   * @return true if set successfully, false otherwise
+   * @throws TachyonException
+   */
+  private boolean setPermission(TachyonURI path, short permission, boolean recursive) throws
+      TachyonException {
+    synchronized (mInodeTree) {
+      checkOwner(path);
+      long opTimeMs = System.currentTimeMillis();
+      Inode targetInode = mInodeTree.getInodeByPath(path);
+      if (recursive && targetInode.isDirectory()) {
+        List<Inode> inodeChildren =
+            mInodeTree.getInodeChildrenRecursive((InodeDirectory) targetInode);
+        for (Inode inode : inodeChildren) {
+          checkOwner(mInodeTree.getPath(inode));
+        }
+        for (Inode inode : inodeChildren) {
+          setAclInternal(inode, opTimeMs, null, null, permission);
+        }
+      }
+      setAclInternal(targetInode, opTimeMs, null, null, permission);
+    }
+    return true;
+  }
+
+  private void setAclInternal(Inode inode, long opTimeMs, String user, String group,
+      short permission) throws TachyonException {
+    // throws exception if security is not enabled.
+    if (!SecurityUtils.isSecurityEnabled(MasterContext.getConf())) {
+      throw new AccessControlException(ExceptionMessage.SECURITY_IS_NOT_ENABLED.getMessage());
+    }
+
+    SetAclEntry.Builder setAcl = SetAclEntry.newBuilder().setId(inode.getId())
+        .setOpTimeMs(opTimeMs);
+    inode.setLastModificationTimeMs(opTimeMs);
+    if (user != null) {
```

Comment:
You can extract the setting into a helper method, since it's also used by `setAclFromEntry`.